File: dotnet/src/HtmlSanitizer.cs
// The idiomatic C# surface over the HtmlSanitizer engine.
//
// Carries no sanitizer logic — every member here marshals to an
// aether_hs_embed_* call in Native.cs.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace HtmlSanitization;

/// <summary>
/// A DOM attribute, <b>borrowed</b> for the duration of a callback.
/// </summary>
/// <remarks>
/// Do not retain one past the callback that gave it to you — the DOM is freed
/// when <see cref="HtmlSanitizer.Sanitize"/> returns.
/// </remarks>
public readonly struct Attribute
{
    internal Attribute(IntPtr ptr) => Pointer = ptr;

    /// <summary>The raw borrowed pointer, for advanced interop.</summary>
    public IntPtr Pointer { get; }

    /// <summary>The attribute's name.</summary>
    public string Name => Native.TakeString(Native.AttrName(Pointer));

    /// <summary>The attribute's value.</summary>
    public string Value => Native.TakeString(Native.AttrValue(Pointer));

    /// <summary>
    /// Rewrite the value in place (e.g. to canonicalise a URL rather than
    /// remove the attribute). The engine copies the string, so the transient
    /// buffer is safe.
    /// </summary>
    public void SetValue(string value) =>
        Native.AttrSetValue(Pointer, Native.Encode(value));

    /// <inheritdoc/>
    public override string ToString() => $"Attribute({Name}={Value})";
}

/// <summary>A DOM node, <b>borrowed</b> for the duration of a callback.</summary>
public readonly struct Node
{
    internal Node(IntPtr ptr) => Pointer = ptr;

    /// <summary>The raw borrowed pointer, for advanced interop.</summary>
    public IntPtr Pointer { get; }

    /// <summary>Document, Element, Text or Comment.</summary>
    public NodeKind Kind => (NodeKind)Native.NodeKindOf(Pointer);

    /// <summary>Element tag name, lowercased by the parser; "" for non-elements.</summary>
    public string Name => Native.TakeString(Native.NodeName(Pointer));

    /// <summary>Text/comment content; "" for elements and documents.</summary>
    public string Value => Native.TakeString(Native.NodeValue(Pointer));

    /// <summary>The parent node, or null at the root.</summary>
    public Node? Parent
    {
        get
        {
            var p = Native.NodeParent(Pointer);
            return p == IntPtr.Zero ? null : new Node(p);
        }
    }

    /// <summary>How many child nodes this node has.</summary>
    public int ChildCount => Native.NodeChildCount(Pointer);

    /// <summary>The child at <paramref name="index"/>.</summary>
    public Node ChildAt(int index) => new(Native.NodeChildAt(Pointer, index));

    /// <summary>Every child node.</summary>
    public IReadOnlyList<Node> Children
    {
        get
        {
            int n = ChildCount;
            var list = new List<Node>(n);
            for (int i = 0; i < n; i++) list.Add(ChildAt(i));
            return list;
        }
    }

    /// <summary>How many attributes this element has.</summary>
    public int AttributeCount => Native.NodeAttrCount(Pointer);

    /// <summary>The attribute at <paramref name="index"/>.</summary>
    public Attribute AttributeAt(int index) => new(Native.NodeAttrAt(Pointer, index));

    /// <summary>Every attribute.</summary>
    public IReadOnlyList<Attribute> Attributes
    {
        get
        {
            int n = AttributeCount;
            var list = new List<Attribute>(n);
            for (int i = 0; i < n; i++) list.Add(AttributeAt(i));
            return list;
        }
    }

    /// <inheritdoc/>
    public override string ToString() => $"Node(Kind={Kind}, Name={Name})";
}

/// <summary>
/// A set-like view over one of the engine's six policy lists. Every operation
/// reads or writes the engine's own set — there is no managed mirror to fall
/// out of sync.
/// </summary>
public sealed class AllowList : IReadOnlyCollection<string>
{
    private readonly HtmlSanitizer _owner;
    private readonly int _which;

    internal AllowList(HtmlSanitizer owner, AllowListKind which)
    {
        _owner = owner;
        _which = (int)which;
    }

    /// <summary>Add one item. Returns this, so calls chain.</summary>
    public AllowList Add(string item)
    {
        Native.Allow(_owner.Handle, _which, Native.Encode(item));
        return this;
    }

    /// <summary>Add several items.</summary>
    public AllowList Add(params string[] items)
    {
        foreach (var i in items) Add(i);
        return this;
    }

    /// <summary>Add every item of a sequence.</summary>
    public AllowList AddRange(IEnumerable<string> items)
    {
        foreach (var i in items) Add(i);
        return this;
    }

    /// <summary>Remove one item — the "deny" direction.</summary>
    public AllowList Remove(string item)
    {
        Native.Disallow(_owner.Handle, _which, Native.Encode(item));
        return this;
    }

    /// <summary>Empty the list — the "start from nothing" move.</summary>
    public AllowList Clear()
    {
        Native.Clear(_owner.Handle, _which);
        return this;
    }

    /// <summary>Is <paramref name="item"/> currently allowed?</summary>
    public bool Contains(string item) =>
        Native.IsAllowed(_owner.Handle, _which, Native.Encode(item)) != 0;

    /// <inheritdoc/>
    public int Count => Native.Count(_owner.Handle, _which);

    /// <summary>The item at <paramref name="index"/>, or "" when out of range.</summary>
    public string this[int index] =>
        Native.TakeString(Native.ItemAt(_owner.Handle, _which, index));

    /// <inheritdoc/>
    public IEnumerator<string> GetEnumerator()
    {
        _owner.ThrowIfDisposed();
        int n = Count;
        for (int i = 0; i < n; i++) yield return this[i];
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    /// <summary>The items, sorted — the deterministic enumeration.</summary>
    public List<string> ToSortedList()
    {
        var list = new List<string>(this);
        list.Sort(StringComparer.Ordinal);
        return list;
    }

    /// <inheritdoc/>
    public override string ToString() => "{" + string.Join(", ", ToSortedList()) + "}";
}

/// <summary>Called before a disallowed tag is removed; return true to KEEP it.</summary>
public delegate bool RemovingTagHandler(Node node, RemovalReason reason);

/// <summary>Called before a disallowed attribute is removed; return true to KEEP it.</summary>
public delegate bool RemovingAttributeHandler(Node element, Attribute attribute, RemovalReason reason);

/// <summary>Called before a disallowed CSS property is removed; return true to KEEP it.</summary>
public delegate bool RemovingStyleHandler(Node element, string name, string value, RemovalReason reason);

/// <summary>Called before a comment is removed; return true to KEEP it.</summary>
public delegate bool RemovingCommentHandler(Node node);

/// <summary>Called for each node (or the document) after filtering.</summary>
public delegate void PostProcessHandler(Node node);

/// <summary>
/// Called for each URL-bearing attribute. Return the URL to use — the
/// <c>resolved</c> argument unchanged for no rewrite, or "" to drop the
/// attribute.
/// </summary>
public delegate string FilterUrlHandler(Node element, string raw, string resolved);

/// <summary>
/// Cleans HTML of constructs that can lead to Cross-Site Scripting (XSS).
/// </summary>
/// <remarks>
/// <code>
/// using var s = new HtmlSanitizer();
/// s.AllowedTags.Add("my-widget");
/// var clean = s.Sanitize("&lt;div onclick=\"evil()\"&gt;hi&lt;/div&gt;");
/// </code>
/// A sanitizer is <b>not</b> thread-safe — the native handle carries mutable
/// policy and hook state. Use one per thread, or guard it.
/// </remarks>
public sealed class HtmlSanitizer : IDisposable
{
    private IntPtr _handle;

    // Registered delegates must be kept alive for as long as the engine can
    // call them. A local delegate would be collected — or its marshalling stub
    // freed — and the process would crash on the next callback. This list is
    // the .NET equivalent of ctypes' keepalive list, and is cleared only in
    // Dispose, after aether_hs_embed_free has run.
    private readonly List<Delegate> _keepAlive = new();

    /// <summary>
    /// Create a sanitizer with the engine's secure defaults populated.
    /// </summary>
    /// <param name="nativeLibraryPath">
    /// An explicit engine path. Otherwise: $HTMLSANITIZER_LIB, then native/
    /// next to the assembly, then ../core/native/, then the OS loader.
    /// </param>
    public HtmlSanitizer(string? nativeLibraryPath = null)
    {
        Native.EnsureResolver(nativeLibraryPath);
        _handle = Native.New();
        if (_handle == IntPtr.Zero)
            throw new InvalidOperationException("failed to create the native sanitizer");

        AllowedTags = new AllowList(this, AllowListKind.Tags);
        AllowedAttributes = new AllowList(this, AllowListKind.Attributes);
        AllowedCssProperties = new AllowList(this, AllowListKind.CssProperties);
        AllowedSchemes = new AllowList(this, AllowListKind.Schemes);
        AllowedClasses = new AllowList(this, AllowListKind.Classes);
        UriAttributes = new AllowList(this, AllowListKind.UriAttributes);
    }

    internal IntPtr Handle
    {
        get
        {
            ThrowIfDisposed();
            return _handle;
        }
    }

    internal void ThrowIfDisposed()
    {
        if (_handle == IntPtr.Zero)
            throw new ObjectDisposedException(nameof(HtmlSanitizer));
    }

    /// <summary>The engine's allowed tag names.</summary>
    public AllowList AllowedTags { get; }

    /// <summary>The engine's allowed attribute names.</summary>
    public AllowList AllowedAttributes { get; }

    /// <summary>The engine's allowed CSS property names.</summary>
    public AllowList AllowedCssProperties { get; }

    /// <summary>The engine's allowed URL schemes.</summary>
    public AllowList AllowedSchemes { get; }

    /// <summary>The engine's allowed CSS class names.</summary>
    public AllowList AllowedClasses { get; }

    /// <summary>Which attributes the engine treats as carrying a URL.</summary>
    public AllowList UriAttributes { get; }

    /// <summary>Where the engine was loaded from, once known.</summary>
    public static string? NativeLibraryPath => Native.ResolvedPath;

    /// <summary>The engine's ABI revision.</summary>
    public static int AbiVersion
    {
        get
        {
            Native.EnsureResolver();
            return Native.AbiVersion();
        }
    }

    // ---- lifecycle ----

    /// <summary>Release the native handle. Idempotent.</summary>
    public void Dispose()
    {
        Release(disposing: true);
        GC.SuppressFinalize(this);
    }

    /// <summary>
    /// Backstop for a dropped sanitizer; <see cref="Dispose"/> is the way.
    /// </summary>
    ~HtmlSanitizer() => Release(disposing: false);

    /// <summary>
    /// The standard Dispose(bool) split. `disposing` is false when this runs
    /// from the finalizer, where touching another managed object is unsafe —
    /// it may already have been finalized. Freeing the native handle is always
    /// safe and is what actually matters; the keepalive list only needs
    /// clearing on the deterministic path, and if we are being finalized the
    /// GC is about to reclaim it regardless.
    /// </summary>
    private void Release(bool disposing)
    {
        if (_handle == IntPtr.Zero) return;
        var h = _handle;
        _handle = IntPtr.Zero;
        Native.Free(h);
        if (disposing)
        {
            // Only now is it certain the engine can no longer invoke a hook.
            _keepAlive.Clear();
        }
    }

    // ---- the main entry point ----

    /// <summary>
    /// Sanitize an HTML fragment. <paramref name="baseUrl"/> resolves relative
    /// URLs; pass "" (the default) for no resolution.
    /// </summary>
    public string Sanitize(string html, string baseUrl = "") =>
        Native.TakeString(Native.Sanitize(Handle, Native.Encode(html), Native.Encode(baseUrl)));

    /// <summary>Sanitize a full HTML document.</summary>
    public string SanitizeDocument(string html, string baseUrl = "") =>
        Native.TakeString(Native.SanitizeDocument(Handle, Native.Encode(html), Native.Encode(baseUrl)));

    // ---- flags ----

    /// <summary>Keep the children of a removed element instead of dropping the subtree.</summary>
    public bool KeepChildNodes
    {
        get => Native.GetKeepChildNodes(Handle) != 0;
        set => Native.SetKeepChildNodes(Handle, value ? 1 : 0);
    }

    /// <summary>Allow <c>data-*</c> attributes through without listing each one.</summary>
    public bool AllowDataAttributes
    {
        get => Native.GetAllowDataAttributes(Handle) != 0;
        set => Native.SetAllowDataAttributes(Handle, value ? 1 : 0);
    }

    // ---- callbacks ----
    //
    // Each On* returns this, so they chain; passing null clears the hook.
    // For the Removing* family, returning true from your handler CANCELS the
    // removal (keeps the node/attribute/property).

    /// <summary>
    /// Hand a trampoline (or null, to clear) to one of the ABI's seven
    /// <c>on_*</c> setters. The trampoline is added to the keepalive list
    /// BEFORE it is registered, so there is no window in which the engine
    /// holds a pointer to a collectable delegate.
    /// </summary>
    private HtmlSanitizer Register(
        Action<IntPtr, IntPtr, IntPtr> register,
        Delegate? trampoline)
    {
        ThrowIfDisposed();
        if (trampoline is null)
        {
            register(_handle, IntPtr.Zero, IntPtr.Zero);
            return this;
        }

        _keepAlive.Add(trampoline);
        // user_data is unused on the .NET side: the delegate already closes
        // over the handler, so there is nothing to look up. The engine's
        // trampoline still round-trips it.
        register(_handle, Marshal.GetFunctionPointerForDelegate(trampoline), IntPtr.Zero);
        return this;
    }

    /// <summary>Called before a disallowed tag is removed; return true to keep it.</summary>
    public HtmlSanitizer OnRemovingTag(RemovingTagHandler? handler)
    {
        if (handler is null) return Register(Native.OnRemovingTag, null);
        CbRemovingTag cb =
            (_, node, reason) => handler(new Node(node), (RemovalReason)reason) ? 1 : 0;
        return Register(Native.OnRemovingTag, cb);
    }

    /// <summary>Called before a disallowed attribute is removed; return true to keep it.</summary>
    public HtmlSanitizer OnRemovingAttribute(RemovingAttributeHandler? handler)
    {
        if (handler is null) return Register(Native.OnRemovingAttribute, null);
        CbRemovingAttribute cb =
            (_, elem, attr, reason) =>
                handler(new Node(elem), new Attribute(attr), (RemovalReason)reason) ? 1 : 0;
        return Register(Native.OnRemovingAttribute, cb);
    }

    /// <summary>Called before a disallowed CSS property is removed; return true to keep it.</summary>
    public HtmlSanitizer OnRemovingStyle(RemovingStyleHandler? handler)
    {
        if (handler is null) return Register(Native.OnRemovingStyle, null);
        CbRemovingStyle cb =
            (_, elem, name, value, reason) =>
                handler(new Node(elem), Native.BorrowString(name),
                        Native.BorrowString(value), (RemovalReason)reason) ? 1 : 0;
        return Register(Native.OnRemovingStyle, cb);
    }

    /// <summary>Called before a comment is removed; return true to keep it.</summary>
    public HtmlSanitizer OnRemovingComment(RemovingCommentHandler? handler)
    {
        if (handler is null) return Register(Native.OnRemovingComment, null);
        CbRemovingComment cb = (_, node) => handler(new Node(node)) ? 1 : 0;
        return Register(Native.OnRemovingComment, cb);
    }

    /// <summary>Called for each node after it has been filtered.</summary>
    public HtmlSanitizer OnPostProcessNode(PostProcessHandler? handler)
    {
        if (handler is null) return Register(Native.OnPostProcessNode, null);
        CbPostProcess cb = (_, node) => handler(new Node(node));
        return Register(Native.OnPostProcessNode, cb);
    }

    /// <summary>Called once with the whole document after filtering.</summary>
    public HtmlSanitizer OnPostProcessDom(PostProcessHandler? handler)
    {
        if (handler is null) return Register(Native.OnPostProcessDom, null);
        CbPostProcess cb = (_, doc) => handler(new Node(doc));
        return Register(Native.OnPostProcessDom, cb);
    }

    /// <summary>
    /// Called for each URL-bearing attribute. The returned string is copied
    /// into a malloc'd C buffer the engine takes ownership of — you do not
    /// free it.
    /// </summary>
    public HtmlSanitizer OnFilterUrl(FilterUrlHandler? handler)
    {
        if (handler is null) return Register(Native.OnFilterUrl, null);
        CbFilterUrl cb = (_, elem, raw, resolved) => Native.StrDup(
            handler(new Node(elem), Native.BorrowString(raw),
                    Native.BorrowString(resolved)));
        return Register(Native.OnFilterUrl, cb);
    }

    // ---- one-shots ----

    /// <summary>Sanitize <paramref name="html"/> with the engine's defaults.</summary>
    public static string SanitizeOnce(string html, string baseUrl = "")
    {
        using var s = new HtmlSanitizer();
        return s.Sanitize(html, baseUrl);
    }

    /// <summary>Sanitize a full document with the engine's defaults.</summary>
    public static string SanitizeDocumentOnce(string html, string baseUrl = "")
    {
        using var s = new HtmlSanitizer();
        return s.SanitizeDocument(html, baseUrl);
    }
}

File: lua/src/htmlsanitizer.lua
--- Clean HTML of constructs that can lead to Cross-Site Scripting (XSS).
---
--- The idiomatic Lua surface over the HtmlSanitizer core. Carries no
--- sanitizer logic — every function here marshals to the C extension in
--- `htmlsanitizer_native` (lua/src/htmlsanitizer.c), which in turn calls the
--- `aether_hs_embed_*` ABI. One sanitizer core, one set of behaviours, N language
--- surfaces.
---
---     local hs = require("htmlsanitizer")
---     local s = hs.new()
---     print(s:sanitize('<div onclick="alert(1)">Hello</div>'))
---     -- <div>Hello</div>
---     s:close()

local native = require("htmlsanitizer_native")

local M = {}

-- ---- ABI constants, re-exported ----

--- Allow-list selectors.
M.TAGS           = native.TAGS
M.ATTRIBUTES     = native.ATTRIBUTES
M.CSS_PROPERTIES = native.CSS_PROPERTIES
M.SCHEMES        = native.SCHEMES
M.CLASSES        = native.CLASSES
M.URI_ATTRIBUTES = native.URI_ATTRIBUTES

--- Why the sanitizer core is about to remove something.
M.REASON_NOT_ALLOWED_TAG       = native.REASON_NOT_ALLOWED_TAG
M.REASON_NOT_ALLOWED_ATTRIBUTE = native.REASON_NOT_ALLOWED_ATTRIBUTE
M.REASON_NOT_ALLOWED_STYLE     = native.REASON_NOT_ALLOWED_STYLE
M.REASON_NOT_ALLOWED_URL_VALUE = native.REASON_NOT_ALLOWED_URL_VALUE
M.REASON_NOT_ALLOWED_VALUE     = native.REASON_NOT_ALLOWED_VALUE
M.REASON_NOT_ALLOWED_CSS_CLASS = native.REASON_NOT_ALLOWED_CSS_CLASS
M.REASON_CLASS_ATTRIBUTE_EMPTY = native.REASON_CLASS_ATTRIBUTE_EMPTY
M.REASON_STYLE_ATTRIBUTE_EMPTY = native.REASON_STYLE_ATTRIBUTE_EMPTY

--- Node kinds.
M.NODE_DOCUMENT = native.NODE_DOCUMENT
M.NODE_ELEMENT  = native.NODE_ELEMENT
M.NODE_TEXT     = native.NODE_TEXT
M.NODE_COMMENT  = native.NODE_COMMENT

-- ---- the allow-list view ----

--- A set-like view over one of the sanitizer core's six policy lists. Every operation
--- reads or writes the sanitizer core's own set — there is no Lua mirror to fall out
--- of sync.
local AllowList = {}
AllowList.__index = AllowList

local function new_allow_list(sanitizer, which)
  return setmetatable({ _s = sanitizer, _which = which }, AllowList)
end

--- Add one item (or every item of a table). Returns self, so calls chain.
function AllowList:add(item)
  if type(item) == "table" then
    for _, v in ipairs(item) do
      self._s._native:allow(self._which, v)
    end
  else
    self._s._native:allow(self._which, item)
  end
  return self
end

--- Remove one item — the "deny" direction. Returns self.
function AllowList:remove(item)
  self._s._native:disallow(self._which, item)
  return self
end

--- Empty the list. Returns self.
function AllowList:clear()
  self._s._native:clear(self._which)
  return self
end

--- Is `item` currently allowed?
function AllowList:contains(item)
  return self._s._native:is_allowed(self._which, item)
end

--- How many entries.
function AllowList:count()
  return self._s._native:count(self._which)
end

AllowList.__len = AllowList.count

--- The item at a 1-based `index`, or "" when out of range. (The ABI is
--- 0-based; the C extension does the conversion so Lua stays 1-based.)
function AllowList:at(index)
  return self._s._native:item_at(self._which, index)
end

--- The items, as a table, in the sanitizer core's own (unspecified but stable) order.
function AllowList:items()
  local out = {}
  for i = 1, self:count() do
    out[i] = self:at(i)
  end
  return out
end

--- The items, sorted — the deterministic version of `items()`.
function AllowList:sorted()
  local out = self:items()
  table.sort(out)
  return out
end

--- Iterate the items: `for item in list:iter() do ... end`
function AllowList:iter()
  local items, i = self:items(), 0
  return function()
    i = i + 1
    return items[i]
  end
end

function AllowList:__tostring()
  return "{" .. table.concat(self:sorted(), ", ") .. "}"
end

-- ---- node / attribute wrappers ----
--
-- The C extension already exposes Node and Attribute as userdata with methods
-- (`node:name()`, `attr:value()`), and stamps each with a generation counter
-- so a retained one errors rather than reading freed memory. They are handed
-- to callbacks unchanged — there is nothing useful for Lua to add, and
-- wrapping would only add a layer that could outlive the borrow.
--
-- Two conveniences that do not retain anything:

--- Every child of `node`, as a table.
function M.children(node)
  local out = {}
  for i = 1, node:child_count() do
    out[i] = node:child_at(i)
  end
  return out
end

--- Every attribute of `node`, as a table.
function M.attributes(node)
  local out = {}
  for i = 1, node:attr_count() do
    out[i] = node:attr_at(i)
  end
  return out
end

-- ---- the sanitizer ----

local Sanitizer = {}
Sanitizer.__index = Sanitizer

--- Create a sanitizer with the sanitizer core's secure defaults populated.
---
--- `native_lib` optionally overrides the library search, which otherwise is:
--- $HTMLSANITIZER_LIB, then `native/` and `../core/native/`, then the OS
--- loader's own path.
function M.new(native_lib)
  local self = setmetatable({}, Sanitizer)
  self._native = native.new(native_lib)
  self._closed = false

  self.allowed_tags           = new_allow_list(self, M.TAGS)
  self.allowed_attributes     = new_allow_list(self, M.ATTRIBUTES)
  self.allowed_css_properties = new_allow_list(self, M.CSS_PROPERTIES)
  self.allowed_schemes        = new_allow_list(self, M.SCHEMES)
  self.allowed_classes        = new_allow_list(self, M.CLASSES)
  self.uri_attributes         = new_allow_list(self, M.URI_ATTRIBUTES)
  return self
end

--- Release the native handle. Idempotent; the userdata's __gc is a backstop,
--- but closing deterministically is better.
function Sanitizer:close()
  if not self._closed then
    self._closed = true
    self._native:close()
  end
end

--- Run `body(s)` with a fresh sanitizer, closing it afterwards. Errors
--- propagate after the close.
function M.use(body, native_lib)
  local s = M.new(native_lib)
  local ok, res = pcall(body, s)
  s:close()
  if not ok then error(res, 0) end
  return res
end

--- Sanitize an HTML fragment. `base_url` resolves relative URLs; omit it (or
--- pass "") for no resolution.
function Sanitizer:sanitize(html, base_url)
  return self._native:sanitize(html, base_url or "")
end

--- Sanitize a full HTML document.
function Sanitizer:sanitize_document(html, base_url)
  return self._native:sanitize_document(html, base_url or "")
end

-- ---- flags ----

--- Keep the children of a removed element instead of dropping the subtree.
function Sanitizer:set_keep_child_nodes(on)
  self._native:set_keep_child_nodes(on and true or false)
  return self
end

function Sanitizer:get_keep_child_nodes()
  return self._native:get_keep_child_nodes()
end

--- Allow `data-*` attributes through without listing each one.
function Sanitizer:set_allow_data_attributes(on)
  self._native:set_allow_data_attributes(on and true or false)
  return self
end

function Sanitizer:get_allow_data_attributes()
  return self._native:get_allow_data_attributes()
end

--- The sanitizer core's ABI revision.
function Sanitizer:abi_version()
  return native.abi_version()
end

-- ---- callbacks ----
--
-- Each `on_*` takes a Lua function (or nil to clear the hook) and returns
-- self, so they chain. For the `removing_*` family, returning a truthy value
-- from your handler CANCELS the removal (keeps the node/attribute/property).
--
-- Handlers are anchored in the Lua registry by the C extension for as long as
-- the sanitizer core can call them, so a local function passed here is safe from
-- collection.

--- `handler(node, reason)` — return true to KEEP the tag.
function Sanitizer:on_removing_tag(handler)
  self._native:on_removing_tag(handler)
  return self
end

--- `handler(elem, attr, reason)` — return true to KEEP the attribute.
function Sanitizer:on_removing_attribute(handler)
  self._native:on_removing_attribute(handler)
  return self
end

--- `handler(elem, name, value, reason)` — return true to KEEP the property.
function Sanitizer:on_removing_style(handler)
  self._native:on_removing_style(handler)
  return self
end

--- `handler(node)` — return true to KEEP the comment.
function Sanitizer:on_removing_comment(handler)
  self._native:on_removing_comment(handler)
  return self
end

--- `handler(node)` — called for each node after it has been filtered.
function Sanitizer:on_post_process_node(handler)
  self._native:on_post_process_node(handler)
  return self
end

--- `handler(doc)` — called once with the whole document after filtering.
function Sanitizer:on_post_process_dom(handler)
  self._native:on_post_process_dom(handler)
  return self
end

--- `handler(elem, raw, resolved) -> string` — return the URL to use;
--- `resolved` unchanged for no rewrite, `""` to drop the attribute.
---
--- The returned string is copied into a malloc'd C buffer the sanitizer core takes
--- ownership of; you do not free it.
function Sanitizer:on_filter_url(handler)
  self._native:on_filter_url(handler)
  return self
end

-- ---- introspection / one-shots ----

--- The sanitizer core's ABI revision, without needing a sanitizer.
function M.abi_version()
  return native.abi_version()
end

--- Where the sanitizer core `.so` was actually loaded from.
function M.engine_path()
  return native.engine_path()
end

--- One-shot: sanitize `html` with the sanitizer core's defaults.
function M.sanitize(html, base_url)
  return M.use(function(s) return s:sanitize(html, base_url) end)
end

--- One-shot: sanitize `html` as a full document, with the sanitizer core's defaults.
function M.sanitize_document(html, base_url)
  return M.use(function(s) return s:sanitize_document(html, base_url) end)
end

M.Sanitizer = Sanitizer
M.AllowList = AllowList

return M
